### PR TITLE
[Client encryption]: Adds support for Brotli compression

### PR DESCRIPTION
**Pull request by JanHyka**
# Pull Request Template

## Description

- Add Brotli compression for .NET8.0 target
- To stabilize performance tests duration results, GC was reconfigured to server mode (reduces StdDev 42% -> 5%)

Brotli compression is now available for .NET8.0 target. All encrypted fields matching criteria are compressed before encryption.

The option is configurable via `Microsoft.Azure.Cosmos.Encryption.Custom.EncryptionOptions` property `CompressionOptions`

`CompressionOptions` structure:
- `Algorithm` : `CompressionOptions.CompressionAlgoritm` suports `None` and `Brotli` (for .NET8.0+), defaults to None.
- `CompressionLevel` : `System.IO.Compression.CompressionLevel` - defaults to CompressionLevel.Fastest
- `MinimalCompressedLength` : determines minimal size of encrypted serialized Json text for compression. Defaults to 128 bytes.

It is suggested to benchmark your use cases, different settings can lead to different ratio of compute vs storage costs. Default settings in general should save both cpu time and memory consumption.

**Warn: make sure all your instances are upgraded to version supporting compression BEFORE enabling it.**

To be processed after #4766 

## Type of change

Please delete options that are not relevant.

- [] New feature (non-breaking change which adds functionality)

## Closing issues

Contributes to #4678

### CODE REVIEW

File: Microsoft.Azure.Cosmos.Encryption.Custom/src/Transformation/MdeEncryptionProcessor.Preview.cs [hunk 1]
```diff
@@ -121,11 +140,26 @@ internal async Task<DecryptionContext> DecryptObjectAsync(
                     continue;
                 }
 
-                (byte[] plainText, int decryptedCount) = this.Encryptor.Decrypt(encryptionKey, cipherTextWithTypeMarker, arrayPoolManager);
+                (byte[] bytes, int processedBytes) = this.Encryptor.Decrypt(encryptionKey, cipherTextWithTypeMarker, arrayPoolManager);
+
+#if NET8_0_OR_GREATER
+                if (encryptionProperties.EncryptionFormatVersion == 4)
+                {
+                    if (encryptionProperties.CompressedEncryptedPaths?.TryGetValue(path, out int decompressedSize) == true)
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos.Encryption.Custom/src/Transformation/MdeEncryptionProcessor.Preview.cs:148]: Algorithm validation/assertion?

File: Microsoft.Azure.Cosmos.Encryption.Custom/src/Transformation/BrotliCompressor.cs [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+                CompressionLevel.Fastest => 1,
+                CompressionLevel.Optimal => 4,
+                CompressionLevel.SmallestSize => 11,
+                _ => throw new ArgumentException("Unsupported compression level", nameof(compressionLevel))
+            };
+        }
+
+        public virtual (byte[], int) Compress(EncryptionProperties properties, string path, byte[] bytes, int length, ArrayPoolManager arrayPoolManager, int compressionLevel)
+        {
+            byte[] compressedBytes = arrayPoolManager.Rent(BrotliEncoder.GetMaxCompressedLength(length));
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos.Encryption.Custom/src/Transformation/BrotliCompressor.cs:30]: In consistency between compression and decompress. 
Compression also not re-sizing and caller is giving the hint of the size.

File: Microsoft.Azure.Cosmos.Encryption.Custom/src/Transformation/BrotliCompressor.cs [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+            if (!BrotliEncoder.TryCompress(bytes.AsSpan(0, length), compressedBytes, out int bytesWritten, compressionLevel, BrotliCompressor.DefaultWindow))
+            {
+                throw new InvalidOperationException();
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos.Encryption.Custom/src/Transformation/BrotliCompressor.cs:34]: Any exception context to carry for troubleshooting?

File: Microsoft.Azure.Cosmos.Encryption.Custom/src/Transformation/BrotliCompressor.cs [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        private const int DefaultWindow = 22;
+
+        internal static int GetQualityFromCompressionLevel(CompressionLevel compressionLevel)
+        {
+            return compressionLevel switch
+            {
+                CompressionLevel.NoCompression => 0,
+                CompressionLevel.Fastest => 1,
+                CompressionLevel.Optimal => 4,
+                CompressionLevel.SmallestSize => 11,
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos.Encryption.Custom/src/Transformation/BrotliCompressor.cs:23]: Based on experimentation on our side Or what will influence the numbers?

File: Microsoft.Azure.Cosmos.Encryption.Custom/src/Transformation/BrotliCompressor.cs [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+namespace Microsoft.Azure.Cosmos.Encryption.Custom.Transformation
+{
+#if ENCRYPTION_CUSTOM_PREVIEW && NET8_0_OR_GREATER
+
+    using System;
+    using System.IO.Compression;
+
+    internal class BrotliCompressor
+    {
+        private const int DefaultWindow = 22;
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos.Encryption.Custom/src/Transformation/BrotliCompressor.cs:14]: Curious, can you please share what the impact higher window number?

Comment (JanHyka) [reply Microsoft.Azure.Cosmos.Encryption.Custom/src/Transformation/BrotliCompressor.cs:14]: I haven't done any experimentation with window tbh, the values are directly taken from
https://github.com/dotnet/runtime/blob/main/src/libraries/System.IO.Compression.Brotli/src/System/IO/Compression/BrotliUtils.cs

File: Microsoft.Azure.Cosmos.Encryption.Custom/src/Transformation/BrotliCompressor.cs (hunk 4 above)
Comment (JanHyka) [reply Microsoft.Azure.Cosmos.Encryption.Custom/src/Transformation/BrotliCompressor.cs:23]: These are taken from https://github.com/dotnet/runtime/blob/9ceb2bfd35ad8edd0342d2b10fef2860ad86ac9b/src/libraries/System.IO.Compression.Brotli/src/System/IO/Compression/BrotliUtils.cs#L15

However it's internal so I had to mirror it.

File: Microsoft.Azure.Cosmos.Encryption.Custom/src/Transformation/BrotliCompressor.cs (hunk 3 above)
Comment (JanHyka) [reply Microsoft.Azure.Cosmos.Encryption.Custom/src/Transformation/BrotliCompressor.cs:34]: I can try with .Compress/.Decompress instead, it seems to be a little bit more communicative about issue encountered (returns status InvalidData/Done/NeedMoreData/DestinationTooSmall).

File: Microsoft.Azure.Cosmos.Encryption.Custom/src/Transformation/MdeEncryptionProcessor.Preview.cs (hunk 1 above)
Comment (JanHyka) [reply Microsoft.Azure.Cosmos.Encryption.Custom/src/Transformation/MdeEncryptionProcessor.Preview.cs:169]: fixed

File: Microsoft.Azure.Cosmos.Encryption.Custom/src/Transformation/BrotliCompressor.cs (hunk 2 above)
Comment (JanHyka) [reply Microsoft.Azure.Cosmos.Encryption.Custom/src/Transformation/BrotliCompressor.cs]: Whole class was refactored, Compress/Decompress are now more symmetric.

File: Microsoft.Azure.Cosmos.Encryption.Custom/src/Transformation/BrotliCompressor.cs (hunk 3 above)
Comment (JanHyka) [reply Microsoft.Azure.Cosmos.Encryption.Custom/src/Transformation/BrotliCompressor.cs]: Updated with non-static method solution.

File: Microsoft.Azure.Cosmos.Encryption.Custom/src/Transformation/BrotliCompressor.cs (hunk 5 above)
Comment (JanHyka) [reply Microsoft.Azure.Cosmos.Encryption.Custom/src/Transformation/BrotliCompressor.cs:15]: I've tried with window size = 10 (minimum allowed)... it led to significantly worse compression (serialization allocation went up 40%, deserialization went up 45%) while both directions took 15% longer... It seems window size 22 is a sweet spot, at least for fastest mode.

**Comment by kirankumarkolli** [comment]
Can you please update the description with usage details (will help consumers)

**Review by kirankumarkolli — DISMISSED** [review]

File: Microsoft.Azure.Cosmos.Encryption.Custom/src/CompressionOptions.cs [hunk 6]
```diff
@@ -33,11 +33,7 @@ public enum CompressionAlgorithm
         /// <summary>
         /// Gets or sets compression algorithm.
         /// </summary>
-#if NET8_0_OR_GREATER
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos.Encryption.Custom/src/CompressionOptions.cs]: Thank you! I missed it.

Comment (kirankumarkolli) [reply Microsoft.Azure.Cosmos.Encryption.Custom/src/CompressionOptions.cs]: Also in future an explicit contract file (like we do for core SDK would be quite useful to pop them-up). I will track it for future.

**Review by kirankumarkolli — APPROVED** [review]

**Pull request merged**